util/tracing/tracer_span.go

```diff
@@ -0,0 +1,568 @@
+	s.LogFields(otlog.String("event", event), otlog.Object("payload", payload))
+}
+
+// Log is part of the opentracing.Span interface. Deprecated.
+func (s *span) Log(data opentracing.LogData) {
+	panic("unimplemented")
+}
+
+// spanGroup keeps track of all the spans that are being recorded as a group (i.e.
+// the span for which recording was enabled and all direct or indirect child
+// spans since then).
+type spanGroup struct {
+	sync.Mutex
+	// spans keeps track of all the local spans. A span is inserted in this slice
+	// as soon as it is opened; the first element is the span passed to
+	// StartRecording().
+	spans []*span
+}
+
+func (ss *spanGroup) addSpan(s *span) {
+	ss.Lock()
+	ss.spans = append(ss.spans, s)
+	ss.Unlock()
+}
+
+// RecordedSpan is just an alias of span which can be used outside the current package.
+type RecordedSpan struct {
+	// ID of the trace; spans that are part of the same hierarchy share
+	// the same trace ID.
+	TraceID uint64
+	// ID of the span.
+	SpanID uint64
+	// Span ID of the parent span.
+	ParentSpanID uint64
+	// Operation name.
+	Operation string
+	// Baggage items get passed from parent to child spans (even through gRPC).
+	// Notably, snowball tracing uses a special `sb` baggage item.
+	Baggage map[string]string
+	// Tags associated with the span.
+	Tags map[string]string
+	// Time when the span was started.
+	StartTime time.Time
+	// Duration in nanoseconds; 0 if the span is not finished.
+	Duration time.Duration
+	// Events logged in the span.
+	Logs []opentracing.LogRecord
+}
+
+// getSpans returns all the local and remote spans accumulated in this group.
+// The first result is the first local span - i.e. the span originally passed to
+// StartRecording().
+func (ss *spanGroup) getSpans() []RecordedSpan {
+	ss.Lock()
+	spans := ss.spans
+	ss.Unlock()
+
+	result := make([]RecordedSpan, 0, len(spans))
+	for _, s := range spans {
+		s.mu.Lock()
+		rs := RecordedSpan{
+			TraceID:      s.TraceID,
+			SpanID:       s.SpanID,
+			ParentSpanID: s.parentSpanID,
+			Operation:    s.operation,
+			StartTime:    s.startTime,
+			Duration:     s.mu.duration,
+		}
+		switch rs.Duration {
+		case -1:
+			// -1 indicates an unfinished span.
+			rs.Duration = 0
+		case 0:
+			// 0 is a special value for unfinished spans. Change to 1ns.
+			rs.Duration = time.Nanosecond
+		}
+
+		if len(s.mu.Baggage) > 0 {
+			rs.Baggage = make(map[string]string)
+			for k, v := range s.mu.Baggage {
+				rs.Baggage[k] = v
+			}
+		}
+		if len(s.mu.tags) > 0 {
+			rs.Tags = make(map[string]string)
+			for k, v := range s.mu.tags {
+				// We encode the tag values as strings.
+				rs.Tags[k] = fmt.Sprint(v)
+			}
+		}
+		rs.Logs = s.mu.recordedLogs
+		s.mu.Unlock()
+		result = append(result, rs)
+	}
+	return result
+}
+
+type noopSpanContext struct{}
+
+var _ opentracing.SpanContext = noopSpanContext{}
+
+func (n noopSpanContext) ForeachBaggageItem(handler func(k, v string) bool) {}
+
+type noopSpan struct {
```

Comment:
I remember opentracing already provide a noopSpan implementation ?

Reply:
It is not exported for outside user.